lib/charms/opensearch/v0/opensearch_users.py

```diff
@@ -251,17 +254,33 @@ def remove_users_and_roles(self, departed_relation_id: Optional[int] = None):
         """
         if not self.opensearch.is_node_up() or not self.unit.is_leader():
             return
+        rel_id = str(departed_relation_id)
 
-        relations = self.model.relations.get(ClientRelationName, [])
-        relation_users = set(
-            [
-                f"{ClientRelationName}_{relation.id}"
-                for relation in relations
-                if relation.id != departed_relation_id
-            ]
-        )
-        self._remove_lingering_users(relation_users)
-        self._remove_lingering_roles(relation_users)
+        relation_users = json.loads(self.charm.peers_data.get(Scope.APP, ClientUsersDict) or "{}")
+        relation_roles = json.loads(self.charm.peers_data.get(Scope.APP, ClientRolesDict) or "{}")
+
+        for username in relation_users.get(rel_id, []):
+            try:
+                self.remove_user(username)
+            except OpenSearchUserMgmtError:
+                logger.error(f"failed to remove user {username}")
+        if rel_id in relation_users:
+            del relation_users[rel_id]
+
+        roles_to_remove = relation_roles.get(rel_id, [])
+        if rel_id in relation_roles:
+            del relation_roles[rel_id]
+        for role in roles_to_remove:
+            # Check if this role is not present in any other relation:
+            if any(role in r for r in relation_roles.values()):
+                continue
+            try:
+                self.remove_role(role)
+            except OpenSearchUserMgmtError:
+                logger.error(f"failed to remove role {role}")
+
+        self.charm.peers_data.put(Scope.APP, ClientUsersDict, json.dumps(relation_users))
+        self.charm.peers_data.put(Scope.APP, ClientRolesDict, json.dumps(relation_roles))
```

Comment:
let's move this to under the `except OpenSearchUserMgmtError` blocks in the deleted methods below.

Reply:
All re-worked, closing thread.